kernel/hashlib.h

```diff
@@ -382,11 +382,19 @@ class commutative_hash {
 	commutative_hash() {
 		buckets.fill(0);
 	}
-	void eat(Hasher h) {
+	template <typename T>
+	void eat(const T &obj) {
+		eat(hash_ops<T>::hash(obj));
+	}
+	void eat(const Hasher &h) {
 		Hasher::hash_t v = h.yield();
 		size_t index = v & (buckets.size() - 1);
 		buckets[index] += v;
 	}
+	template <typename T, typename U>
+	void eat_pair(const T &v, const U &u) {
```

Comment:
Surely there is no need for adding eat_pair? There is a way to have support for eating any sort of hashable container (pair, tuple, vector, array, what have you) and that is not adding eat_pair, eat_tuple, etc, it's what we're doing already

Reply:
The problem is, if you have a reference to a T and a U separately, and you want to compute the hash of the pair (T, U), building a `std::pair<T, U>` requires copying the T and U into the `std::pair`, which is unnecessarily expensive when T and U are complex (e.g. contain an `IdString` or `std::vector`). In `OptMergePass` this situation occurs [here](https://github.com/YosysHQ/yosys/blob/b95549b469a4bc4664818109572ad722de73be51/passes/opt/opt_merge.cc#L110) and [here](https://github.com/YosysHQ/yosys/blob/b95549b469a4bc4664818109572ad722de73be51/passes/opt/opt_merge.cc#L54).

Reply:
The latter case might not be important because the result of the `extract()` call should (or could) be moved into the `std::pair` so there's no extra copy there.

Reply:
I see, that's an important thing to know about the purposes of the PR, I thought you were only going for clean code here. How do you feel about this approach?

```diff
+               using pair_hasher = hash_ops<std::pair<const SigBit&, const SigSpec&>>;
                for (int i = 0; i < s_width; i++)
-                       comm.eat(hash_ops<std::pair<SigBit, SigSpec>>::hash({sig_s[i], sig_b.extract(i*width, width)}));
+                       comm.eat(pair_hasher::hash({sig_s[i], sig_b.extract(i*width, width)}));
```

Reply:
I'd still prefer to infer the type names. What do you think about my latest attempt?